File: ArrayAndHashing/EncodeAndDecodeStrings/eds.py
from typing import List


class Solution:

    def encode(self, strs: List[str]) -> str:
        parts = []
        for s in strs:
            parts.append(f"{len(s)}#{s}")
        return "".join(parts)

    def decode(self, s: str) -> List[str]:
        res = []
        i = 0
        bck = 0
        while i < len(s):
            if s[i] == "#":
                n = int(s[bck:i])
                ss = s[i + 1 : i + n + 1]
                res.append(ss)
                i = i + n + 1
                bck = i
            else:
                i += 1
        return res


if __name__ == "__main__":
    sol = Solution()

    n = ["neet", "code", "love", "you"]
    e = sol.encode(n)
    print(f"TE1: {e}")
    d = sol.decode(e)
    print(f"TD1: {d}")

    n = ["we", "say", ":", "yes"]
    e = sol.encode(n)
    print(f"TE1: {e}")
    d = sol.decode(e)
    print(f"TD1: {d}")

File: ArrayAndHashing/EncodeAndDecodeStrings/eds.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func encode(strs []string) string {
	var sb strings.Builder
	for _, s := range strs {
		sb.WriteString(fmt.Sprintf("%d#%s", len(s), s))
	}
	return sb.String()
}

func decode(str string) []string {
	var res []string

	i := 0
	back := 0
	for i < len(str) {
		if str[i] == '#' {
			n, _ := strconv.Atoi(str[back:i])
			res = append(res, str[i+1:i+n+1])
			i = i + n + 1
			back = i
		}
		i++
	}

	return res
}

func main() {
	s1 := []string{"neet", "code", "love", "you"}
	e := encode(s1)
	fmt.Printf("E1: %s\n", e)
	fmt.Printf("D1: %v\n", decode(e))

	s2 := []string{"we", "say", ":", "yes"}
	e2 := encode(s2)
	fmt.Printf("E1: %s\n", e2)
	fmt.Printf("D1: %v\n", decode(e2))
}
